File: Scratchpad/Scratchpad.c
#include <stdio.h>
#include <stdlib.h>
#include <Block.h>

#define b Block_copy
#define r return

typedef struct OrderLine {
    int id;
    double price;
} OrderLine;

typedef struct Order {
    int id;
    OrderLine* orderLines;
    int lineCount;
} Order;

typedef void* (^ (^ (^NthElement)(int))(void*))(int);
typedef OrderLine* (^ (^ (^FindOrderLine)(int))(OrderLine*))(int);
typedef void (^ (^ (^ (^ReplaceOrderLine)(int))(OrderLine))(OrderLine*))(int);
typedef Order (^ (^ (^ChangeOrderLinePrice)(Order))(int))(double);

NthElement nthElement = ^(int n) { r b(^(void* list) { r b(^(int listCount) {
            r (n < listCount) ? (void*)&((char*)list)[n * sizeof(list[0])] : (void*)NULL; }); }); };

FindOrderLine findOrderLine = ^(int orderLineId) { r b(^(OrderLine* orderLines) { r b(^(int lineCount) {
            for (int i = 0; i < lineCount; ++i) {
                if (orderLines[i].id == orderLineId) {
                    r &orderLines[i];
                }
            }
            r (OrderLine*)NULL; }); }); };

ReplaceOrderLine replaceOrderLine = (^(int orderLineId) { r b(^(OrderLine newOrderLine) { r b(^(OrderLine* orderLines) { r b(^(int lineCount) {
                for (int i = 0; i < lineCount; ++i) {
                    if (orderLines[i].id == orderLineId) {
                        orderLines[i] = newOrderLine;
                        break;
                    } } }); }); }); });

ChangeOrderLinePrice changeOrderLinePrice = ^(Order order) { r b(^(int orderLineId) { r b(^(double newPrice) {
            OrderLine* orderLine = findOrderLine(orderLineId)(order.orderLines)(order.lineCount);
            if (orderLine != NULL) {
                OrderLine newOrderLine = { .id = orderLine->id, .price = newPrice };
                replaceOrderLine(orderLineId)(newOrderLine)(order.orderLines)(order.lineCount);
            }
            Order newOrder = { .id = order.id, .orderLines = order.orderLines, .lineCount = order.lineCount };
            r newOrder; }); }); };

int main() {
    Order order = {
      .id = 1, 
      .orderLines = (OrderLine[]){{.id = 1, .price = 10.0}, {.id = 2, .price = 20.0}}, .lineCount = 2};

    Order updatedOrder = changeOrderLinePrice(order)(1)(15.0);

    OrderLine* orderLine = nthElement(0)(updatedOrder.orderLines)(updatedOrder.lineCount);
    if (orderLine) {
        printf("OrderLine ID: %d, Price: %.2f\n", orderLine->id, orderLine->price);
    } else {
        printf("The list does not have an element at that index\n");
    }
}

File: Scratchpad/Scratchpad.py
from typing import List, Callable, Optional, Any

class OrderLine:
    def __init__(self, id: int, price: float):
        self.id = id
        self.price = price

class Order:
    def __init__(self, id: int, order_lines: List[OrderLine]):
        self.id = id
        self.order_lines = order_lines

NthElement = Callable[[int], Callable[[List[Any]], Optional[Any]]]

FindOrderLine = Callable[[int], Callable[[List[OrderLine]], Optional[OrderLine]]]

ReplaceOrderLine = Callable[[int, OrderLine], Callable[[List[OrderLine]], List[OrderLine]]]

ChangeOrderLinePrice = Callable[[Order], Callable[[int], Callable[[float], Order]]]

nth_element: NthElement = lambda n: lambda list: (
    list[n] if n < len(list) else None)

find_order_line: FindOrderLine = lambda order_line_id: lambda order_lines: (
    next((ol for ol in order_lines if ol.id == order_line_id), None ))

replace_order_line: ReplaceOrderLine = lambda order_line_id, new_order_line: lambda order_lines: (
    [new_order_line if ol.id == order_line_id else ol for ol in order_lines ])

change_order_line_price: ChangeOrderLinePrice = lambda order: lambda order_line_id: lambda new_price: (
    Order(
        order.id,
        replace_order_line(order_line_id, OrderLine(order_line_id, new_price))(
            order.order_lines
    )))

order = Order(
    1,
    [OrderLine(1, 10.0), OrderLine(2, 20.0)])

updated_order = change_order_line_price(order)(1)(15.0)

order_line = nth_element(0)(updated_order.order_lines)
print(
    order_line.price
    if order_line
    else "The list does not have an element at that index"
)
